lib/model/message.dart

```diff
@@ -1,19 +1,310 @@
+    required String? realmEmptyTopicDisplayName,
+  }) async {
+    final localMessageId = _nextLocalMessageId++;
+    assert(!outboxMessages.containsKey(localMessageId));
+
+    final now = (ZulipBinding.instance.utcNow().millisecondsSinceEpoch / 1000).toInt();
+    _outboxMessages[localMessageId] = switch (destination) {
+      StreamDestination(:final streamId, :final topic) => StreamOutboxMessage(
+        localMessageId: localMessageId,
+        selfUserId: selfUserId,
+        timestamp: now,
+        conversation: StreamConversation(
+          streamId,
+          topic.interpretAsServer(
+            // Because either of the values can get updated, the actual topic
+            // can change, for example, between "(no topic)" and "general chat",
+            // or between different names of "general chat".  This should be
+            // uncommon during the lifespan of an outbox message.
+            //
+            // There's also an unavoidable race that has the same effect:
+            // an admin could change the name of "general chat"
+            // (i.e. the value of realmEmptyTopicDisplayName) concurrently with
+            // the user making the send request, so that the setting in effect
+            // by the time the request arrives is different from the setting the
+            // client last heard about.  The realm update events do not have
+            // information about this race for us to update the prediction
+            // correctly.
+            zulipFeatureLevel: zulipFeatureLevel,
+            realmEmptyTopicDisplayName: realmEmptyTopicDisplayName),
+          displayRecipient: null),
+        content: content),
+      DmDestination(:final userIds) => DmOutboxMessage(
+        localMessageId: localMessageId,
+        selfUserId: selfUserId,
+        timestamp: now,
+        conversation: DmConversation(allRecipientIds: userIds),
+        content: content),
+    };
+
+    _outboxMessageDebounceTimers[localMessageId] = Timer(kLocalEchoDebounceDuration, () {
+      assert(outboxMessages.containsKey(localMessageId));
+      _outboxMessageDebounceTimers.remove(localMessageId);
+      _updateOutboxMessage(localMessageId, newState: OutboxMessageState.waiting);
+    });
+
+    _outboxMessageWaitPeriodTimers[localMessageId] = Timer(kSendMessageRetryWaitPeriod, () {
+      assert(outboxMessages.containsKey(localMessageId));
+      _outboxMessageWaitPeriodTimers.remove(localMessageId);
+      _updateOutboxMessage(localMessageId, newState: OutboxMessageState.waitPeriodExpired);
+    });
+
+    try {
+      await _apiSendMessage(connection,
+        destination: destination,
+        content: content,
+        readBySender: true,
+        queueId: queueId,
+        localId: localMessageId.toString());
+    } catch (e) {
+      // `localMessageId` is not necessarily in the store. This is because
+      // message event can still arrive before the send request fails to
+      // networking issues.
```

Comment:
nit: maybe "fails with networking issues" or "fails because of networking issues"